### PR TITLE
Fatafat280

### PR DESCRIPTION
**Pull request by ligadata-william**
@ligadata-alsion Please find below the documentation I've written for this utility.

@ligadata-dan please review my code and take it for a spin to see how it works.

Documentation for CleanUtil-1.0

The goal of CleanUtil is to provide a tool that will allow a user to clean up an environment being used under test. The tools capabilities extend to cleaning topics in kafka, deleting of the node base path being used by Kamanja in zookeeper and the dropping of tables from Metadata and Datastore in the database.

***** IMPORTANT *****
Please note this tool is NOT recommended for use in production environments. It is designed purely as a test tool and misuse of the tool can cause permanent loss of data.
***** IMPORTANT *****

How it Works:

CleanUtil should be presented the same configuration file as is given to MetadataAPI for performing CRUD operations on metadata. This file must be passed in command line. You may find an example file at $KAMANJA_HOME/config/MetadataAPIConfig.properties. CleanUtil will take this properties file and use it to determine the database being used to store your cluster configuration (Alison, link to cluster configuration documentation here please). It will then read in the cluster configuration you have saved to ascertain the remaining data stores' store types, schema names and locations. It will also pull the zookeeper and input/output adapter configurations. From here, it will take the user operations to determine what (if any) parts of the environment to clean. Using this method, CleanUtil will be able to clean a clustered environment so long as the location the user is running this utility from has network access to the hosts on the environment.

Usage:

Prior to using, please turn off any running nodes of KamanjaManager but leave on all zookeeper nodes, kafka brokers and databases.

To use CleanUtil-1.0, simply navigate to $KAMANJA_HOME/bin where the utility should be installed to.

The usage is as shown here:

CleanUtil --config /path/to/MetadataAPIConfig.properties [--clean-kafka] [--clean-zookeeper] [--clean-testdata [List of messages/containers]] [--clean-metadata] [--cleanstatusinfo]
    
or 

CleanUtil --config /path/to/MetadataAPIConfig.properties [--clean-all [List of messages/containers]

Explanation of options:

--config: --config be given or it will fail immediately. The configuration file given should be the same configuration given when running any kamanja (MetadataAPI) commands ($KAMANJA_HOME/config/MetadataAPIConfig.properties is used by Kamanja by default). This tool will NOT default to any configuration file.

--clean-kafka: a switch indicating CleanUtil should delete the kafka topics found in cluster configuration. When present in the command, it will pull the names of all topics from the adapter configuration and make calls using the Kafka API to delete the topics.

Example: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-kafka

--clean-zookeeper: a switch indicating CleanUtil should delete the ZooKeeperNodeBasePath from zookeeper found in cluster configuration. When presented as part of the command, it will pull the Zookeeper configuration from cluster configuration, read the ZooKeeperNodeBasePath and delete it from zookeeper.

Example: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-zookeeper

--clean-metadata: a switch indicating CleanUtil should delete all metadata from the database. It will determine the location and store type given in cluster configuration and proceed to drop all tables from its schema name. This includes metadata, model configuration and cluster configuration. NOTE: If you run --clean-metadata, you will need to add cluster configuration to metadata prior to running this tool again.

Example: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-metadata

--clean-testdata: a switch indicating CleanUtil should delete all testdata from the database except for persisted messages and containers. It will remove all tables related to the kafka adapter.

Example1: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-testdata

This switch also takes an optional comma-separated list of messages and/or containers. These messages/containers are then transformed into their associated table names and removed from the database. For instance, if you have message history stored and the namespace of the message is 'com.ligadata.messages' and the name of the message is 'customer' and you also have container information stored with namespace 'com.ligadata.containers' and name 'purchasehistory', you would run:

Example2: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-testdata com.ligadata.messages.customer,com.ligadata.containers.purchasehistory

and it would delete the standard tables from testdata as well as the specified message/container data from the database.

--clean-all: a switch that indicates that all aspects of the environment should be cleaned. Running:

Example: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-all

is the same as running:

Example: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-kafka --clean-zookeeper --clean-metadata --clean-testdata

It also supported optional messages/containers as an argument to delete specific tables from the DataStore.

Example: CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-kafka com.ligadata.messages.customer,com.ligadata.containers.purchasehistory

Further Examples:

1. If you wish to clean your environment but retain your metadata and persisted messages and containers:

CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-kafka --clean-zookeeper --clean-testdata

2. To clean your environment and delete persisted messages and containers while retaining your metadata:

CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-kafka --clean-zookeeper --clean-testdata com.ligadata.messages.customer,com.ligadata.containers.purchasehistory

3. To delete data from kafka but retain metadata and persisted messages and containers:

CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-testdata

or

CleanUtil-1.0 --config $KAMANJA_HOME/config/MetadataAPIConfig.properties --clean-testdata com.ligadata.messages.customer,com.ligadata.containers.purchasehistory

In both examples, the kafka adapter offset will be reset back to 0 upon starting KamanjaManager and it will re-process all messages currently in the input topic. This is good for demos where you want to show off results without having to perform metadata operations or push data again.

If you're output is dependent on message history, the second example shows how to remove the message history to allow for subsequent runs that produce the same output. If you're model does not rely on message history of container information, the first option should be enough.

### CODE REVIEW

**Comment by ligadata-krishna** [comment]
Instead of separate utility, this functionality should be part of kamanja commands such as kamanja clean or something like that. 

 

thanks

**Comment by ligadata-william** [comment]
It can be, but like all Kamanja commands, it will still be the Kamanja
script calling some other utility.

**Comment by ligadata-william** [comment]
Additionally, a bash script isn't enough to perform this functionality. A
coded project was required.

**Comment by ligadata-dhaval** [comment]
Hi,

So, shall I add the utility script to the kamanja command line or do we
need to a coded project always to create the jar?

Regards,
Dhaval Kolapkar

Best Regards,
Dhaval Kolapkar
Platform Engineer Intern | LigaDATA

**Comment by ligadata-krishna** [comment]
From user experience, we should have same type of usage commands and help associated with them. This should also fit in the same framework as other Kamanja commands, utilize the same components and follow same security as well as exception/error handling.

**Comment by ligadata-alison** [comment]
Hi William,
    You requested a link to cluster configuration. I have:

http://kamanja.org/advanced-developers-guide/#Configuring_Clusters

but there is no content in this. Can you help with content? Or give me a
rough idea of what I might put in there?
thanks,
alison

**Comment by ligadata-alison** [comment]
Hi William,
    I'm trying to build content for the Cluster Configuration section. Is
this a good start?

http://kamanja.org/advanced-developers-guide/configuring-clusters/

thanks,
alison

File: trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanKafka.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      logger.info(s"CLEAN-UTIL: Deleting kafka topic '$topicName'")
+      AdminUtils.deleteTopic(zkClient, topicName)
+    }
+    else {
+      logger.warn(s"CLEAN-UTIL: Unable to locate Kafka topic '$topicName'. Skipping delete...")
+      return
+    }
+
+    val timeout = 5.seconds.fromNow
+    while(timeout.hasTimeLeft()) {
```
Comment (ligadata-dan) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanKafka.scala]: Should we sleep for some millisecs?  that way we don't hammer the AdminUtils....

File: trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanStores.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  private val logger = org.apache.log4j.Logger.getLogger(this.getClass)
+
+  def cleanMetadata(dataStore: DataStore): Unit = {
+    logger.info("Dropping tables from metadata...")
+    dataStore.DropContainer(Array("model_config_objects", "metadata_objects", "transaction_id", "jar_store", "config_objects"))
+  }
+
+  def cleanDatastore(dataStore: DataStore, tables: Option[Array[String]]): Unit = {
+    logger.info("Dropping tables from data store...")
+    var tableArr = Array("checkpointinformation", "globalcounters", "modelresults", "adapteruniqkvdata")
```
Comment (ligadata-dan) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanStores.scala]: Nitpicking here, but can we define the table names as constants

File: trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    try{
+      if(args.length == 0) {
+        logger.error("CLEAN-UTIL: No parameters were given.")
+        logger.error(usage)
+        return
+      }
+      else {
+        val options = nextOption(Map(), args.toList)
+        val configFile = options.getOrElse('config, null)
+        config = new CleanerConfiguration(configFile.asInstanceOf[String])
```
Comment (ligadata-dan) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala]: NPE if config is Null.????

Comment (ligadata-dan) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala]: configFile, I mean..

File: trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+          else {
+            CleanStores.cleanDatastore(config.dataStore, Some(cleanTestdata.asInstanceOf[String].split(',')))
+          }
+        }
+      }
+    }
+    catch {
+      case e: MissingArgumentException => logger.error(e)
+      case e: InvalidArgumentException => logger.error(e)
+      case e: Exception => logger.error("Unexpected Exception caught:\n" + e)
```
Comment (ligadata-dan) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala]: Log stacktrace???

File: trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala (hunk 3 above)
Comment (ligadata-krishna) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala]: I just want make it clear - we need make any functionality like this as
Kamanja commands if we are going to ship to customers and use same
framework and security mechanisms.

I ask every check-in to be critically code reviewed by at least one team
member and that process really helps identifying issues earliest possible.
It is best to do code reviews with the person that wrote the code by
walking through it with reviewer.

Comment (ligadata-william) [inline trunk/Utils/CleanUtil/src/main/scala/com/ligadata/utils/clean/CleanUtil.scala]: Dan has actually been doing the code review with me and is simply making comments on github for future reference when I go to make the changes.

**Comment by ligadata-william** [comment]
@ligadata-dan I made several changes to this branch. Please take another look to make sure I didn't miss anything.

I want to make enhancements to the kamanja bash script to give access to this utility, but the way it currently parses paramaters won't work for this. I don't know that we can properly express this in the grammar we desire in the bash script. I'm not sure it's worth the effort to include it in the kamanja script rather than allocate enough time to finish the kamanja shell that Dan was working on before the previous two sprints sidetracked us.

**Comment by ligadata-rich** [comment]
I concur with @ligadata-william on this.  Going forward, the positional
approach currently used in the Kamanja script is not flexible enough for
the integration of the utility (not to mention access to other metadata
that we are really not supporting yet ... ).   A dsl is really needed for
this to allow optional terms, lists/arrays of things, etc.

The ultimate is the REPL thing that Dan was working on that supports a dsl
(or rather multiple dsls) that allow for the these more complex
queries/commands.

**Comment by ligadata-william** [comment]
@ligadata-dan I've added additional functionality to check if any MetadataAPIService or Kamanja nodes are running. If so, the utility should exit without cleaning. Please review.

**Comment by ligadata-william** [comment]
@ligadata-dan, have you had any time to review my changes?

**Comment by ligadata-william** [comment]
I didn't realize this hadn't already been merged into Sprint8Features. I'm closing this and I'll merge it into dev.

**Pull request closed without merging**